pkg/fleet-manager/application_helper.go

```diff
@@ -541,3 +547,57 @@ func generateFleetKey(app *applicationapi.Application) client.ObjectKey {
 		Name:      fleetName,
 	}
 }
+
+// getPolicyDestination returns the actual destination used by the sync policy.
+// The function assumes either Application or its SyncPolicy will have a valid Destination, as this is ensured by the webhook validator.
+// If SyncPolicy's Destination is nil, it defaults to Application's Destination.
+func getPolicyDestination(app *applicationapi.Application, policy *applicationapi.ApplicationSyncPolicy) applicationapi.ApplicationDestination {
+	if policy.Destination == nil {
+		return applicationapi.ApplicationDestination{
+			Fleet:           app.Spec.Destination.Fleet,
+			ClusterSelector: app.Spec.Destination.ClusterSelector,
+		}
+	}
+	return applicationapi.ApplicationDestination{
+		Fleet:           policy.Destination.Fleet,
+		ClusterSelector: policy.Destination.ClusterSelector,
+	}
+}
+
+// doesClusterMatchSelector checks if a cluster's labels match the provided selector.
+func doesClusterMatchSelector(cluster *clusterv1alpha1.Cluster, selector *applicationapi.ClusterSelector) bool {
+	// If there is no selector, all clusters are considered a match.
+	if selector == nil || selector.MatchLabels == nil {
+		return true
+	}
+
+	for key, value := range selector.MatchLabels {
+		if clusterValue, ok := cluster.Labels[key]; !ok || clusterValue != value {
+			// If the cluster does not have a label for the key,
+			// or the cluster's label value does not match the selector's value,
+			// this cluster does not match the selector.
+			return false
+		}
+	}
+
+	return true
+}
+
+// doesAttachedClusterMatchSelector checks if a attachedCluster's labels match the provided selector.
+func doesAttachedClusterMatchSelector(attachedCluster *clusterv1alpha1.AttachedCluster, selector *applicationapi.ClusterSelector) bool {
```

Comment:
I can merge with the above by passing the labels. Or alternatively, we can even use go generic